### ISSUE TITLE
wrong error when using a unit struct as length of a repeat expression

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#27008]
```rust
struct A { i: i32};
let a = [0; A{i: 42}];
```
yields the expected
```
error: expected positive integer for repeat count, found struct [E0306]
```

while 

```rust
struct B;
let b = [0; B];
```

thinks that `B` is a variable
```
error: expected constant integer for repeat count, found variable [E0307]
```

[PlayPen](http://is.gd/IgyXUb)

**Comment by rick68** [comment]
I hope this can help you http://is.gd/NzhhNN

**Comment by Aatch** [comment]
This is due to the fact that while `B` is a path expression, it doesn't refer to a constant, so the relevant code fails, and the check that produces the error assumes that `B` is a variable.

**Comment by Detegr** [comment]
What would be the proper way to fix this (rustc newbie here)? Is it possible to check if a non-constant path expression refers to an integer or to a struct, or should the error message just be changed to contain the two possibilities?

**Comment by oli-obk** [comment]
It's probably necessary to change https://github.com/rust-lang/rust/blob/master/src/librustc/middle/const_eval.rs#L875-935 to account for unit-struct-names
